### PR TITLE
Add a ConnectionListener

### PR DESCRIPTION
**Pull request by yschimke**
Given that we have happy eyeballs, and are considering features like client side load balancing, having visibility into connections seems warranted.

Also specifically for Android, dealing with different networks makes this important.

### CODE REVIEW

File: okhttp/src/jvmTest/java/okhttp3/ConnectionListenerTest.kt [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Tag
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.Timeout
+import org.junit.jupiter.api.extension.RegisterExtension
+
+@Flaky // STDOUT logging enabled for test
+@Timeout(30)
+@Tag("Slow")
+class ConnectionListenerTest {
```
Comment (yschimke) [inline okhttp/src/jvmTest/java/okhttp3/ConnectionListenerTest.kt:42]: This is currently just a subset of EventListenerTest without too much thought of the interesting cases.

File: okhttp/src/jvmMain/kotlin/okhttp3/ConnectionListener.kt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * Listener for connection events. Extend this class to monitor the new connections and closes.
+ *
+ * All event methods must execute fast, without external locking, cannot throw exceptions,
+ * attempt to mutate the event parameters, or be reentrant back into the client.
+ * Any IO - writing to files or network should be done asynchronously.
+ */
+abstract class ConnectionListener {
+  open fun connectionOpening(route: Route) {}
```
Comment (yschimke) [inline okhttp/src/jvmMain/kotlin/okhttp3/ConnectionListener.kt:29]: Comments to come, but for now working out what the interesting events and use cases would be.

- Track the open connections
- Track connection usage, do we keep connection open too long, how many requests on a single connection.
- Route failures causing delays.

Presumably request failures can be tracked via an EventListener already.
-

File: okhttp/src/jvmMain/kotlin/okhttp3/ConnectionListener.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  open fun connectFailed(route: Route, failure: IOException) {}
+  open fun connectionOpened(connection: Connection) {}
+  open fun connectionClosed(connection: Connection) {}
+  open fun connectionAcquired(connection: Connection, call: Call) {}
```
Comment (yschimke) [inline okhttp/src/jvmMain/kotlin/okhttp3/ConnectionListener.kt:33]: TODO - should we identify reasons connections are not eligible? Any parts of RealConnection.isEligible that would not be expected.

File: okhttp/src/jvmMain/kotlin/okhttp3/ConnectionListener.kt (hunk 2 above)
Comment (swankjesse) [reply okhttp/src/jvmMain/kotlin/okhttp3/ConnectionListener.kt:29]: Rename to `connectStart` ?
Include the motivating `Call` ?

File: okhttp/src/jvmMain/kotlin/okhttp3/ConnectionListener.kt [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ * Listener for connection events. Extend this class to monitor the new connections and closes.
+ *
+ * All event methods must execute fast, without external locking, cannot throw exceptions,
+ * attempt to mutate the event parameters, or be reentrant back into the client.
+ * Any IO - writing to files or network should be done asynchronously.
+ */
+abstract class ConnectionListener {
+  open fun connectionOpening(route: Route) {}
+  open fun connectFailed(route: Route, failure: IOException) {}
+  open fun connectionOpened(connection: Connection) {}
```
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/ConnectionListener.kt:31]: `connectEnd`  ? I’m attempting (foolish?) symmetry with EventListener

File: okhttp/src/jvmMain/kotlin/okhttp3/ConnectionListener.kt (hunk 3 above)
Comment (swankjesse) [reply okhttp/src/jvmMain/kotlin/okhttp3/ConnectionListener.kt:33]: I don’t think it’s that interesting?

File: okhttp/src/jvmMain/kotlin/okhttp3/ConnectionListener.kt [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  open fun connectionClosed(connection: Connection) {}
+  open fun connectionAcquired(connection: Connection, call: Call) {}
+  open fun connectionReleased(connection: Connection, call: Call) {}
```
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/ConnectionListener.kt:34]: I wonder if we should have `successCount` here?

Comment (swankjesse) [reply okhttp/src/jvmMain/kotlin/okhttp3/ConnectionListener.kt:34]: Or perhaps something else where we track possibly-non-fatal exceptions, like RealConnection.trackFailure() ?

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnection.kt [hunk 6]
```diff
@@ -90,6 +92,13 @@ class RealConnection(
    * Once true this is always true. Guarded by this.
    */
   var noNewExchanges = false
+    set(value) {
```
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnection.kt:95]: I expect this probably breaks our rules around not calling EventListeners inside a synchronized block?

**Review by swankjesse — COMMENTED** [review]
I think this is a great idea. Let’s ship it in 5.0.

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnection.kt (hunk 6 above)
Comment (swankjesse) [reply okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnection.kt:95]: (but why didn’t the test fail?)

File: okhttp/src/jvmMain/kotlin/okhttp3/ConnectionListener.kt (hunk 5 above)
Comment (yschimke) [reply okhttp/src/jvmMain/kotlin/okhttp3/ConnectionListener.kt:34]: Yep - let's add in a follow up.  We should definitely expose this information here, but worth dedicated focus.

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnection.kt (hunk 6 above)
Comment (yschimke) [reply okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnection.kt:95]: I've added tests that catch this, will fix before landing.

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnection.kt [hunk 7]
```diff
@@ -90,6 +92,15 @@ class RealConnection(
...
+      val previousValue = field
+      field = value
+      if (!previousValue) {
+        queueEvent {
+          connectionListener.noNewExchanges(this)
```
Comment (yschimke) [inline okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnection.kt:100]: This is the one that feels unfortunate... Relies on `connection.synchronizedWithEvents`

File: okhttp/src/jvmMain/kotlin/okhttp3/ConnectionListener.kt (hunk 5 above)
Comment (swankjesse) [reply okhttp/src/jvmMain/kotlin/okhttp3/ConnectionListener.kt:56]: Good call. Maybe it aught to be a property on the connection.

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnection.kt (hunk 7 above)
Comment (swankjesse) [reply okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnection.kt:102]: Yeah I think we should just pull up the event-notifying side-effect into the callsite

**Review by swankjesse — COMMENTED** [review]
I’d like to find a way to avoid the event queueing. . . might not be possible? Lemme take a stab at it.

**Comment by swankjesse** [comment]
I _think_ it’s possible 'cause we already plumbed through the `Socket` so we could close it without holding a lock.

File: okhttp-testing-support/src/main/kotlin/okhttp3/RecordingConnectionListener.kt [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+  override fun connectStart(route: Route, call: Call) = logEvent(ConnectionEvent.ConnectStart(System.nanoTime(), route, call))
+
+  override fun connectFailed(route: Route, call: Call, failure: IOException) = logEvent(ConnectionEvent.ConnectFailed(System.nanoTime(), route, call, failure))
+
+  override fun connectEnd(connection: Connection) {
+    logEvent(ConnectionEvent.ConnectEnd(System.nanoTime(), connection))
+  }
+
+  override fun connectionClosed(connection: Connection) = logEvent(ConnectionEvent.ConnectionClosed(System.nanoTime(), connection))
```
Comment (swankjesse) [inline okhttp-testing-support/src/main/kotlin/okhttp3/RecordingConnectionListener.kt:120]: We should assert that Start & Close events are balanced.

File: okhttp-testing-support/src/main/kotlin/okhttp3/RecordingConnectionListener.kt [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+  override fun connectionAcquired(connection: Connection, call: Call) {
+    logEvent(ConnectionEvent.ConnectionAcquired(System.nanoTime(), connection, call))
+  }
+
+  override fun connectionReleased(connection: Connection, call: Call) {
```
Comment (swankjesse) [inline okhttp-testing-support/src/main/kotlin/okhttp3/RecordingConnectionListener.kt:126]: We should assert that Acquire & Release events are balanced.

File: okhttp/src/jvmMain/kotlin/okhttp3/ConnectionListener.kt [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+  /**
+   * Invoked when a call no longer uses a connection.
+   */
+  open fun connectionReleased(connection: Connection, call: Call) {}
+
+  /**
+   * Invoked when a connection is marked for no new exchanges.
+   */
+  open fun noNewExchanges(connection: Connection) {}
```
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/ConnectionListener.kt:61]: FYI, with my pushed commits we don’t send this event if we’re definitely going to close the connection immediately

File: okhttp/src/jvmTest/java/okhttp3/ConnectionListenerTest.kt [hunk 11]
```diff
@@ -52,11 +52,11 @@ open class ConnectionListenerTest {
   private val listener = RecordingConnectionListener()
   private val handshakeCertificates = localhost()
 
-  open val fastFallback: Boolean = true
+  open val fastFallback: Boolean get() = true
```
Comment (yschimke) [inline okhttp/src/jvmTest/java/okhttp3/ConnectionListenerTest.kt:55]: What's the difference here?

File: okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt [hunk 12]
```diff
@@ -222,6 +223,7 @@ class RouteFailureTest {
       "ConnectionAcquired",
       "NoNewExchanges",
       "ConnectionReleased",
+      "ConnectionClosed",
```
Comment (yschimke) [inline okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt:226]: Nice!

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnectionPool.kt [hunk 13]
```diff
@@ -118,6 +127,7 @@ class RealConnectionPool(
     connection.assertThreadHoldsLock()
 
     connections.add(connection)
+//    connection.queueEvent { connectionListener.connectEnd(connection) }
```
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnectionPool.kt:130]: What to do here? I think we need a test to check symmetry before we can drop this comment

File: okhttp-testing-support/src/main/kotlin/okhttp3/RecordingConnectionListener.kt [hunk 14]
```diff
@@ -95,6 +95,11 @@ open class RecordingConnectionListener : ConnectionListener() {
   }
 
   private fun logEvent(e: ConnectionEvent) {
+    if (e.connection != null) {
```
Comment (yschimke) [inline okhttp-testing-support/src/main/kotlin/okhttp3/RecordingConnectionListener.kt:98]: Nice.

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnectionPool.kt [hunk 15]
```diff
@@ -252,7 +265,6 @@ class RealConnectionPool(
       Platform.get().logCloseableLeak(message, callReference.callStackTrace)
 
       references.removeAt(i)
-      connection.noNewExchanges = true
```
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnectionPool.kt:255]: Very small behavior change here - if we leak a Call that doesn‘t immediately mark the connection with noNewExchanges. That’s fine.

File: okhttp-testing-support/src/main/kotlin/okhttp3/RecordingConnectionListener.kt (hunk 9 above)
Comment (yschimke) [reply okhttp-testing-support/src/main/kotlin/okhttp3/RecordingConnectionListener.kt:126]: I can add. Probably as a fast follow.

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnectionPool.kt (hunk 13 above)
Comment (yschimke) [reply okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnectionPool.kt:130]: OK, I can add, unless you plan to.

File: okhttp/src/jvmTest/java/okhttp3/ConnectionListenerTest.kt (hunk 11 above)
Comment (swankjesse) [reply okhttp/src/jvmTest/java/okhttp3/ConnectionListenerTest.kt:55]: Avoids a potential problem accessing state from the superclass constructor.

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnectionPool.kt (hunk 13 above)
Comment (swankjesse) [reply okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/RealConnectionPool.kt:130]: Lets do follow-up. This PR is big enough!

**Pull request merged**